Implement variable argument function call issue in checked blocks (#251) & modify BOUNDS_CHECKED pragma (#247)

wonsubkim opened the pull request:
+ Prevent variable argument function call in checked blocks (#251)
  + In checked block, it does not allow call to variable argument function

+ Modify implementation BOUNDS_CHECKED pragma (#247)
  + For elaborate scope handling, implementation is changed to use annotation token
  + In previous implementation, BOUNDS_CHECKED pragma is handled in incorrect scope
  When parsing '#pragma BOUNDS_CHECKED', it is processed with both '{' and '}'
  When entering/exiting scope, it is not combined with proper scope
  To bind BOUNDS_CHECKED pragma with accurate scope, it generates annotation token for pragma
  When consuming annotation token(tok::annot_pragma_bounds_checked), it can recognize scope corresponding to pragma
  Annotation token pragma is also used in some other pragma handling
  + #pragma BOUNDS_CHECKED [ON/OFF/DEFAULT]
  + corresponding tests are also commited

wonsubkim commented:
BOUNDS_CHECKED pragma (#247) error case was found by extending checked_scope_pragma.c
checked scope property is sometimes incorrectly set because BOUNDS_CHECKED pragma is processed with brace token. Before parsing brace token, it is processed together.
For elaborate process of pragma, I have generated annotation token after parsing pragma
This annotation token is independently parsed after l_brace or r_brace
Therefore, we can handle accurate scope property

dtarditi commented on include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -8928,6 +8928,9 @@ def err_bounds_type_annotation_lost_checking : Error<
     "function without a prototype cannot be used or declared in a checked scope">;
+
+  def err_checked_scope_no_variadic_func : Error<
+    "call to variable argument function cannot be used in a checked scope">;
```
I suggest changing the wording from "cannot be used" to "cannot be made"

dtarditi commented on lib/Parse/ParsePragma.cpp:
```diff
@@ -2204,39 +2214,22 @@ void PragmaMSIntrinsicHandler::HandlePragma(Preprocessor &PP,
 // Handle the checked-c top level scope checked property.
-// #pragma BOUNDS_CHECKED on
-// #pragma BOUNDS_CHECKED off
+// #pragma BOUNDS_CHECKED [on-off-swtich]
```
Typo: swtich -> switch.

dtarditi commented on lib/Sema/SemaExpr.cpp:
```diff
@@ -394,11 +394,18 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, SourceLocation Loc,
   // type in checked block.
   // Otherwise, it does not emit redundant error message at the use of variable
   // since it already produced an error message at the declaration of variable.
+  // Calls to variable argument functions are not allowed in a checked block.
```
You can declare a function pointer that takes variable arguments and has a variable argument type.   I think you need check this at the type level too, just as you a check for uses of unchecked pointer types.

This makes me wonder about what we should do when you have constructed types in a checked block that use unchecked pointer types.  For example, you could have a pointer to an unchecked pointer.    Checked code could never use the unchecked pointer, but it could pass it around.

dtarditi replied:
I opened a Checked C spec issue about constructed types that use unchecked pointer types: https://github.com/Microsoft/checkedc/issues/142.   Feel free to comment on it there.

wonsubkim commented:
For constructed type & variable arguments type, I have modified
I have implemented recursive unchecked type checking & variadic type checking
Related test code is also released

wonsubkim commented:
I am ready for review  :)
